opencog/atomspace/AtomTable.cc

```diff
@@ -498,23 +495,45 @@ Handle AtomTable::add(AtomPtr atom, bool async)
     atom->keep_incoming_set();
     atom->setAtomSpace(_as);
 
+    ContentHash ch = atom->get_hash();
+
+    // Lock, to prevent two different threads from trying to add exactly
+    // the same atom.
+    std::unique_lock<std::recursive_mutex> lck(_store_mtx);
+
+    // Search again, now that we hold the lock. This is necessary for
+    // the case where another atom was just inserted by another
+    // thread while we were cloning and copying values in this thread.
+    auto range = _atom_store.equal_range(ch);
+    auto bkt = range.first;
+    auto end = range.second;
+    for (; bkt != end; bkt++) {
+        if (*((AtomPtr) bkt->second) == *atom) {
+            return bkt->second;
+        }
+    }
+
     _size++;
     if (atom->isNode()) _num_nodes++;
     if (atom->isLink()) _num_links++;
     _size_by_type[atom->_type] ++;
 
     Handle h(atom->getHandle());
-    _atom_store.insert({atom->get_hash(), h});
-
-    if (not _transient and not async)
-        put_atom_into_index(atom);
+    _atom_store.insert({ch, h});
 
-    // We can now unlock, since we are done.
+    // We can now unlock, since we are done updating the store
+    // and counts.
     lck.unlock();
 
-    // Update the indexes asynchronously
-    if (not _transient and async)
-        _index_queue.enqueue(atom);
+    // Note: we don't need to hold the lock for the
+    // index since it locks an index-specific mutex.
+    if (not _transient)
+        {
+        if (async)
```

Comment:
FYI, totally unrelated to your changes --  the "async" design here is broken, and that entire code branch should just be removed.